### PR TITLE
IngressController NodePort publishing strategy

### PR DESCRIPTION
**Pull request by ironcladlou**
Introduce a NodePort publishing strategy for IngressControllers.

/cc @openshift/sig-network-edge @smarterclayton

### CODE REVIEW

File: enhancements/network/ingress-nodeport-publishing.md [hunk 1]
```diff
@@ -0,0 +1,248 @@
+---
+title: ingress-nodeport-publishing
+authors:
+  - "@ironcladlou"
+reviewers:
+  - "@ironcladlou"
+  - "@smarterclayton"
+  - "@knobc"
```
Comment (Miciah) [inline enhancements/network/ingress-nodeport-publishing.md]: You missed a few of the letters in "knobunc" here and below.

File: enhancements/network/ingress-nodeport-publishing.md [hunk 2]
```diff
@@ -109,24 +110,25 @@ spec:
...
+As an OpenShift administrator, I want to integrate the default ingresscontroller with a self-managed load balancer directly through node ports in a way that maximizes ingress utilization and ingresscontroller scheduling flexibility.
+
+### Implementation Details/Notes/Constraints
+
+One critical architectural detail of this proposal which demands scrutiny is the following constraint:
+
+* The Ingress Operator will ignore any updates to `.spec.ports[].nodePort` fields of the Service.
 
-### Implementation Details/Notes/Constraints [optional]
+By making explicit that users own the  `.spec.ports[].nodePort` field, no additional port configuration API should be required. By default, ports are allocated automatically, and if those automatically allocated ports are insufficient, users can update the managed `Service` resource directly.
```
Comment (Miciah) [inline enhancements/network/ingress-nodeport-publishing.md]: What does "insufficient" mean?  Does this refer to the use-case in which the user determines the ports ahead of time (possibly as part of firewall configuration) and then configure the ingresscontroller to use those predetermined ports?

**Comment by Miciah** [comment]
Will `NodePort` become the default for some platforms?  Should the EP address defaults?

File: enhancements/network/ingress-nodeport-publishing.md (hunk 2 above)
Comment (ironcladlou) [reply enhancements/network/ingress-nodeport-publishing.md]: I've tried to clarify this use case.

**Comment by ironcladlou** [comment]
>Will NodePort become the default for some platforms? Should the EP address defaults?

I think NodePort should replace HostNetwork in all defaulting contexts if possible, and so I've added that to the proposal for discussion.

**Comment by ironcladlou** [comment]
The defaulting may turn out to be more nuanced, and it it bogs things down I would suggest extracting it into a separate EP.

**Comment by Miciah** [comment]
Yeah, I'm concerned that changing the default could cause surprise (you deleted and recreated your ingresscontroller without specifying the endpoint publishing strategy, and the default changed with the last upgrade—surprise!) and raises the question of migrations (you installed a 4.4 cluster and upgraded a 4.3 cluster to 4.4, and they behave differently—surprise!).

**Comment by ironcladlou** [comment]
>Yeah, I'm concerned that changing the default could cause surprise (you deleted and recreated your ingresscontroller without specifying the endpoint publishing strategy, and the default changed with the last upgrade—surprise!) and raises the question of migrations (you installed a 4.4 cluster and upgraded a 4.3 cluster to 4.4, and they behave differently—surprise!).

I went ahead and removed the parts about changing the default. It's immaterial to the core feature and seems like it'll only confuse the topic for now.

File: enhancements/network/ingress-nodeport-publishing.md [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+
+### Implementation Details/Notes/Constraints
+
+One critical architectural detail of this proposal which demands scrutiny is the following constraint:
+
+* The Ingress Operator will ignore any updates to `.spec.ports[].nodePort` fields of the Service.
+
+By making explicit that users own the  `.spec.ports[].nodePort` field, no additional port configuration API should be required. By default, ports are allocated automatically and users can discover the actual port allocations for integrations. However, sometimes static port allocations are necessary to integrate with existing infrastructure which may not be easily reconfigured in response to dynamic ports. To achieve integrations with static node ports, users can update the managed `Service` resource directly.
+
+Because OpenShift isn't managing anything connected to the NodePort service, the ports used to expose the IngressController are irrelevant and can be left to the discretion of the administrator (constrained only by the cluster node port CIDR configuration).
```
Comment (smarterclayton) [inline enhancements/network/ingress-nodeport-publishing.md]: Awesome

**Comment by smarterclayton** [comment]
Looks good to me at a high level.

**Comment by ironcladlou** [comment]
@smarterclayton @danehans raised the good question: "Can the [MetalLB](http://metallb.universe.tf) LoadBalancer Service implementation support any of these use cases?"

I haven't yet revisited the topic very deeply but wanted to throw the idea your way in case you already some thoughts as to what overlap there might be.

**Comment by ironcladlou** [comment]
One quick note is that MetalLB itself requires [configuration](https://metallb.universe.tf/configuration/) which seems like it would have to become OpenShift configuration API (e.g. L2 mode, BGP mode, IP pools, etc.)

Maybe that would end up looking something like what we're doing for CoreDNS.

**Comment by smarterclayton** [comment]
MetalLb should support nodeport adequately - Ive asked PM to make sure it’s
in the backlog (it’s a prerequisite for full master management)

On Nov 22, 2019, at 1:39 PM, Dan Mace <notifications@github.com> wrote:

@smarterclayton <https://github.com/smarterclayton> @danehans
<https://github.com/danehans> raised the good question: "Can the MetalLB
<http://metallb.universe.tf> LoadBalancer Service implementation support
any of these use cases?"

I haven't yet revisited the topic very deeply but wanted to throw the idea
your way in case you already some thoughts as to what overlap there might
be.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/openshift/enhancements/pull/99?email_source=notifications&email_token=AAI37J46WJEFL5QL25E7X43QVARPFA5CNFSM4JJFJ5PKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEE6P2LA#issuecomment-557645100>,
or unsubscribe

**Comment by ironcladlou** [comment]
@smarterclayton PTAL, I can open an API PR imminently

**Pull request merged**